File: acm/src/leetCode/repository/No3.java
package leetCode.repository;

import java.util.HashMap;
import java.util.Map;

/**
 * 给定一个字符串，找出不含有重复字符的最长子串的长度。
 * <p>
 * 示例：
 * <p>
 * 给定 "abcabcbb" ，没有重复字符的最长子串是 "abc" ，那么长度就是3。
 * <p>
 * 给定 "bbbbb" ，最长的子串就是 "b" ，长度是1。
 * <p>
 * 给定 "pwwkew" ，最长子串是 "wke" ，长度是3。请注意答案必须是一个子串，"pwke" 是 子序列  而不是子串。
 */
public class No3 {

    /**
     * 查找最长不重复子串
     *
     * @param s 字符串
     * @return 最长不重复子串长度
     */
    public int lengthOfLongestSubstring(String s) {
        return windowSlide(s);
    }

    public int windowSlide(String s) {
        Map<Character, Integer> map = new HashMap<>();

        int max = 0;
        for (int i = 0, j = 0; j < s.length(); j++) {
            Integer next = map.get(s.charAt(j));
//            System.out.println("next:" + next + "  i:" + i + "  j:" + j);
            if (null == next) {
            } else {
                i = Math.max(next + 1, i);
            }
            max = Math.max(max, j - i + 1);
            map.put(s.charAt(j), j);
        }
        return max;
    }

    public static void main(String[] args) {

        System.out.println(new No3().lengthOfLongestSubstring("tmmzuxt"));
    }

}

File: rust/src/leetcode/no_3.rs
use std::collections::HashMap;

/**
    Given a string s, find the length of the longest substring without duplicate characters.

Example 1:

Input: s = "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3.

Example 2:

Input: s = "bbbbb"
Output: 1
Explanation: The answer is "b", with the length of 1.

Example 3:

Input: s = "pwwkew"
Output: 3
Explanation: The answer is "wke", with the length of 3.
Notice that the answer must be a substring, "pwke" is a subsequence and not a substring.
*/
struct Solution;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut c_cnt = HashMap::<char, usize>::new();

        let vec_c: Vec<char> = s.chars().collect();

        let mut max_length = 0;
        let mut last_index = 0;
        for (i, c) in vec_c.iter().enumerate() {
            if c_cnt.contains_key(c) {
                let new_l = c_cnt.get(c).unwrap().clone();
                for j in last_index..=new_l {
                    c_cnt.remove(vec_c.get(j).unwrap());
                }
                last_index = new_l + 1;
            } else {
                max_length = max_length.max(i - last_index + 1);
            }
            c_cnt.insert(*c, i);
        }
        max_length as i32
    }
}

#[test]
fn test() {
    assert_eq!(0, Solution::length_of_longest_substring("".to_string()));
    assert_eq!(1, Solution::length_of_longest_substring(" ".to_string()));
    assert_eq!(
        3,
        Solution::length_of_longest_substring("abcabcbb".to_string())
    );
    assert_eq!(
        1,
        Solution::length_of_longest_substring("bbbbb".to_string())
    );
    assert_eq!(
        3,
        Solution::length_of_longest_substring("pwwkew".to_string())
    );
}
